ssl/quic/quic_impl.c

```diff
@@ -4324,10 +4323,117 @@ SSL *ossl_quic_new_listener_from(SSL *ssl, uint64_t flags)
 /*
  * SSL_new_from_listener
  * ---------------------
+ * code here is derived from ossl_quic_new(). The `ssl` argument is
+ * a listener object which already comes with QUIC port/engine. The newly
+ * created QUIC connection object (QCSO) is going to share the port/engine
+ * with listener (`ssl`).  The `ssl` also becomes a parent of QCSO created
+ * by this function. The caller uses QCSO instance to connect to
+ * remote QUIC server.
+ *
+ * The QCSO created here requires us to also create a channel so we
+ * can connect to remote server.
  */
 SSL *ossl_quic_new_from_listener(SSL *ssl, uint64_t flags)
 {
-    /* TODO(QUIC SERVER): Implement SSL_new_from_listener */
+    QCTX ctx;
+    QUIC_CONNECTION *qc;
+    QUIC_LISTENER *ql;
+    SSL_CONNECTION *sc = NULL;
+
+    if (flags != 0)
+        return NULL;
+
+    if (!expect_quic_listener(ssl, &ctx))
+        return NULL;
+
+    if (!SSL_up_ref(&ctx.ql->obj.ssl))
+        return NULL;
+
+    qctx_lock(&ctx);
```

Comment:
Does this lock need to happen prior the SSL_up_ref, in the event the connection is closed by another thread while this is occuring?

Reply:
I must admit I did not put much thought into this when copying the code  `ossl_quic_new_listener_from()`.

I assume _connection is closed_ here means application is closing/destroying listener QUIC object. To be honest I don't think it matters here. If I understand your concern right here we talk about situation which looks as follows:
```
static SSL *qssl_listener;

static void thread_1(void)
{
    SSL_free(qssl_listener);
    qssl_listener = NULL;
}

static void thread_2(void)
{
    SSL *qssl_conn;

    qssl_qconn = SSL_new_from_listener(qssl_listener);
}
```

There are two threads, thread 1 attempts to close the the listener which is a global variable `qssl_listener`, while thread 2 attempts to use `qssl_listener` to create a new connection object. If it is the case then changing the order of qctx_lock() with SSL_up_ref() won't change anything. We are too low to fix the race between `SSL_free()` and `SSL_new_from_listener()` it's the application itself to make its mind whether listener should be closed or new quic connection object should be created.

or is there something else on your mind?